The type-checker assumes that Trait<'a>::AssociatedType may be outlived by 'a, even with "where AssociatedType: 'a"

trait Trait<'a> {
        type Ty;
        fn method(ty_ref: &'a Self::Ty) where Self::Ty: 'a { }
    }
    
    fn caller<'a, T: Trait<'a>>(arg: &'a T::Ty) where T::Ty: 'a {
        T::method(arg)
    }

Compilation fails on the playground (for stable 1.36.0 and nightly 1.38.0) with this error:

    Compiling playground v0.0.1 (/playground)
    error[E0309]: the associated type `<T as Trait<'_>>::Ty` may not live long enough
     --> src/lib.rs:7:9
      |
    7 |         T::method(arg)
      |         ^^^^^^^^^
      |
      = help: consider adding an explicit lifetime bound `<T as Trait<'_>>::Ty: 'a`...
    note: ...so that the type `<T as Trait<'_>>::Ty` will meet its required lifetime bounds
     --> src/lib.rs:7:9
      |
    7 |         T::method(arg)
      |         ^^^^^^^^^
    
    error: aborting due to previous error

Compilation succeeds if I make any of the following changes:

- Add a lifetime constraint to `Ty` at its definition site, `type Ty: 'a;`

- Move the lifetime parameter from the trait to the method, `fn method<'a>(...)`

Compilation fails with the same error message if I change `method` into a free function:

    trait Trait<'a> {
        type Ty;
    }
    
    fn method<'a, T: Trait<'a>>(_arg: &'a T::Ty) where T::Ty: 'a { }
    
    fn caller<'a, T: Trait<'a>>(arg: &'a T::Ty) where T::Ty: 'a {
        method::<T>(arg)
    }

Even if this is intended behaviour, the error message could be improved.

--- a/tests/ui/associated-types/assoc-type-outlives-via-where-clause-63253.rs
+++ b/tests/ui/associated-types/assoc-type-outlives-via-where-clause-63253.rs
@@ -1,0 +1,33 @@
+//! Regression test for <https://github.com/rust-lang/rust/issues/63253>.
+//!
+//! A `where Self::Ty: 'a` bound on the callee was not being used to prove the
+//! associated type outlives `'a` at the call site, so both of these calls used
+//! to fail with E0309 ("the associated type `<T as Trait<'_>>::Ty` may not live
+//! long enough").
+
+//@ check-pass
+
+#![allow(unused)]
+
+// The associated function is reached through a method-call path.
+trait Trait<'a> {
+    type Ty;
+    fn method(ty_ref: &'a Self::Ty) where Self::Ty: 'a {}
+}
+
+fn caller<'a, T: Trait<'a>>(arg: &'a T::Ty) where T::Ty: 'a {
+    T::method(arg)
+}
+
+// The same bound, reached through a free function instead.
+trait Trait2<'a> {
+    type Ty;
+}
+
+fn free_fn<'a, T: Trait2<'a>>(_arg: &'a T::Ty) where T::Ty: 'a {}
+
+fn free_fn_caller<'a, T: Trait2<'a>>(arg: &'a T::Ty) where T::Ty: 'a {
+    free_fn::<T>(arg)
+}
+
+fn main() {}

